# Activity: Build a Naive Bayes model 

## Introduction

In this activity, you will build your own Naive Bayes model. Naive Bayes models can be valuable to use any time you are doing work with predictions because they give you a way to account for new information. In today's world, where data is constantly evolving, modeling with Naive Bayes can help you adapt quickly and make more accurate predictions about what could occur.

For this activity, you work for a firm that provides insights for management and coaches in the National Basketball Association (NBA), a professional basketball league in North America. The league is interested in retaining players who can last in the high-pressure environment of professional basketball and help the team be successful over time. In the previous activity, you analyzed a subset of data that contained information about the NBA players and their performance records. You conducted feature engineering to determine which features would most effectively predict a player's career duration. You will now use those insights to build a model that predicts whether a player will have an NBA career lasting five years or more. 

The data for this activity consists of performance statistics from each player's rookie year. There are 1,341 observations, and each observation in the data represents a different player in the NBA. Your target variable is a Boolean value that indicates whether a given player will last in the league for five years. Since you previously performed feature engineering on this data, it is now ready for modeling.   

## Step 1: Imports

### Import packages

Begin with your import statements. Of particular note here are `pandas` and from `sklearn`, `naive_bayes`, `model_selection`, and `metrics`.

In [3]:
# Import relevant libraries and modules.

import pandas as pd
from sklearn import naive_bayes, model_selection, metrics

### Load the dataset

Recall that in the lab about feature engineering, you outputted features for the NBA player dataset along with the target variable ``target_5yrs``. Data was imported as a DataFrame called `extracted_data`. As shown in this cell, the dataset has been automatically loaded in for you. You do not need to download the .csv file, or provide more code, in order to access the dataset and proceed with this lab. Please continue with this activity by completing the following instructions.

In [4]:
# RUN THIS CELL TO IMPORT YOUR DATA.
# Load extracted_nba_players.csv into a DataFrame called extracted_data.

extracted_data = pd.read_csv('extracted_nba_players.csv')

### Display the data

Review the first 10 rows of data.

In [5]:
# Display the first 10 rows of data.

### YOUR CODE HERE ###
extracted_data.head(10)


fg    3p    ft  reb  ast  stl  blk  tov  target_5yrs  total_points  \
0  34.7  25.0  69.9  4.1  1.9  0.4  0.4  1.3            0         266.4   
1  29.6  23.5  76.5  2.4  3.7  1.1  0.5  1.6            0         252.0   
2  42.2  24.4  67.0  2.2  1.0  0.5  0.3  1.0            0         384.8   
3  42.6  22.6  68.9  1.9  0.8  0.6  0.1  1.0            1         330.6   
4  52.4   0.0  67.4  2.5  0.3  0.3  0.4  0.8            1         216.0   
5  42.3  32.5  73.2  0.8  1.8  0.4  0.0  0.7            0         277.5   
6  43.5  50.0  81.1  2.0  0.6  0.2  0.1  0.7            1         409.2   
7  41.5  30.0  87.5  1.7  0.2  0.2  0.1  0.7            1         273.6   
8  39.2  23.3  71.4  0.8  2.3  0.3  0.0  1.1            0         156.0   
9  38.3  21.4  67.8  1.1  0.3  0.2  0.0  0.7            0         155.4   

   efficiency  
0    9.722628  
1    9.368030  
2   25.150327  
3   28.500000  
4   18.782609  
5   24.342105  
6   37.541284  
7   26.563107  
8   15.757576  
9   18.282353

## Step 2: Model preparation

### Isolate your target and predictor variables
Separately define the target variable (`target_5yrs`) and the features.

In [6]:
# Define the y (target) variable.

### YOUR CODE HERE ###
y = extracted_data.target_5yrs

# Define the X (predictor) variables.

### YOUR CODE HERE ###
X = extracted_data.drop('target_5yrs', axis=1)

<details>
<summary><h4><strong>Hint 1</strong></h4></summary>

Refer to [the content about splitting your data into X and y](https://www.coursera.org/learn/the-nuts-and-bolts-of-machine-learning/lecture/VxbUT/construct-a-naive-bayes-model-with-python).
</details>

<details>
<summary><h4><strong>Hint 2</strong></h4></summary>

In `pandas`, subset your DataFrame by using square brackets `[]` to specify which column(s) to select.
</details>

<details>
<summary><h4><strong>Hint 3</strong></h4></summary>

Quickly subset a DataFrame to exclude a particular column by using the `drop()` function and specifying the column to drop.
</details>

### Display the first 10 rows of your target data

Display the first 10 rows of your target and predictor variables. This will help you get a sense of how the data is structured.

In [7]:
# Display the first 10 rows of your target data.

### YOUR CODE HERE ###
y.head(10)


0    0
1    0
2    0
3    1
4    1
5    0
6    1
7    1
8    0
9    0
Name: target_5yrs, dtype: int64

**Question:** What do you observe about the your target variable?


[Write your response here. Double-click (or enter) to edit.]

In [8]:
# Display the first 10 rows of your predictor variables.

### YOUR CODE HERE ###
X.head(10)

fg    3p    ft  reb  ast  stl  blk  tov  total_points  efficiency
0  34.7  25.0  69.9  4.1  1.9  0.4  0.4  1.3         266.4    9.722628
1  29.6  23.5  76.5  2.4  3.7  1.1  0.5  1.6         252.0    9.368030
2  42.2  24.4  67.0  2.2  1.0  0.5  0.3  1.0         384.8   25.150327
3  42.6  22.6  68.9  1.9  0.8  0.6  0.1  1.0         330.6   28.500000
4  52.4   0.0  67.4  2.5  0.3  0.3  0.4  0.8         216.0   18.782609
5  42.3  32.5  73.2  0.8  1.8  0.4  0.0  0.7         277.5   24.342105
6  43.5  50.0  81.1  2.0  0.6  0.2  0.1  0.7         409.2   37.541284
7  41.5  30.0  87.5  1.7  0.2  0.2  0.1  0.7         273.6   26.563107
8  39.2  23.3  71.4  0.8  2.3  0.3  0.0  1.1         156.0   15.757576
9  38.3  21.4  67.8  1.1  0.3  0.2  0.0  0.7         155.4   18.282353

**Question:** What do you observe about the your predictor variables?

[Write your response here. Double-click (or enter) to edit.]

### Perform a split operation on your data

Divide your data into a training set (75% of data) and test set (25% of data). This is an important step in the process, as it allows you to reserve a part of the data that the model has not observed. This tests how well the model generalizes—or performs—on new data.

In [9]:
# Perform the split operation on your data.
# Assign the outputs as follows: X_train, X_test, y_train, y_test.

### YOUR CODE HERE ###
X_train, X_test, y_train, y_test = model_selection.train_test_split(X, y, test_size=0.25, stratify=y, random_state=42)

<details>
<summary><h4><strong>Hint 1</strong></h4></summary>

Refer to [the content about splitting your data between a training and test set](https://www.coursera.org/learn/the-nuts-and-bolts-of-machine-learning/lecture/VxbUT/construct-a-naive-bayes-model-with-python).
</details>

<details>
<summary><h4><strong>Hint 2</strong></h4></summary>

Call the function in the `model_selection` module of `sklearn` on the features and target variable, in order to perform the splitting.
</details>

<details>
<summary><h4><strong>Hint 3</strong></h4></summary>

Call the `model_selection.train_test_split()` function, passing in both `features` and `target`, while configuring the appropriate `test_size`.

Assign the output of this split as `X_train`, `X_test`, `y_train`, `y_test`.
</details>

### Print the shape of each output 

Print the shape of each output from your train-test split. This will verify that the split operated as expected.

In [10]:
# Print the shape (rows, columns) of the output from the train-test split.

# Print the shape of X_train.

### YOUR CODE HERE ###
print(X_train.shape)


# Print the shape of X_test.

### YOUR CODE HERE ###
print(X_test.shape)


# Print the shape of y_train.

### YOUR CODE HERE ###
print(y_train.shape)


# Print the shape of y_test.

### YOUR CODE HERE ###
print(y_test.shape)


(1005, 10)
(335, 10)
(1005,)
(335,)


<details>
<summary><h4><strong>Hint 1</strong></h4></summary>

Call the attribute that DataFrames in `pandas` have to get the number of rows and number of columns as a tuple.
</details>

<details>
<summary><h4><strong>Hint 2</strong></h4></summary>

Call the `shape` attribute.
</details>

**Question:** How many rows are in each of the outputs?


[Write your response here. Double-click (or enter) to edit.]

**Question:** What was the effect of the train-test split?


[Write your response here. Double-click (or enter) to edit.]

## Step 3: Model building

**Question:** Which Naive Bayes algorithm should you use?

[Write your response here. Double-click (or enter) to edit.]

<details>
<summary><h4><strong>Hint 1</strong></h4></summary>

Refer to [the content about different implementations of the Naive Bayes](https://www.coursera.org/learn/the-nuts-and-bolts-of-machine-learning/supplement/1zfDy/naive-bayes-classifiers) to determine which is appropriate in this situation.
</details>

<details>
<summary><h4><strong>Hint 2</strong></h4></summary>

Note that you are performing binary classification.
</details>

<details>
<summary><h4><strong>Hint 3</strong></h4></summary>

You can identify the appropriate algorithm to use because you are performing a binary classification and assuming that the features of your model follow a normal distribution.
</details>

### Fit your model to your training data and predict on your test data

By creating your model, you will be drawing on your feature engineering work by training the classifier on the `X_train` DataFrame. You will use this to predict `target_5yrs` from `y_train`.

Start by defining `nb` to be the relevant algorithm from `sklearn`.`naive_bayes`. Then fit your model to your training data. Use this fitted model to create predictions for your test data.

In [12]:
# Assign `nb` to be the appropriate implementation of Naive Bayes.

### YOUR CODE HERE ###
nb = naive_bayes.GaussianNB()


# Fit the model on your training data.

### YOUR CODE HERE ###
nb.fit(X_train, y_train)


# Apply your model to predict on your test data. Call this "y_pred".

### YOUR CODE HERE ###
y_pred = nb.predict(X_test)


<details>
<summary><h4><strong>Hint 1</strong></h4></summary>

Refer to [the content about constructing a Naive Bayes](https://www.coursera.org/learn/the-nuts-and-bolts-of-machine-learning/lecture/VxbUT/construct-a-naive-bayes-model-with-python).
</details>

<details>
<summary><h4><strong>Hint 2</strong></h4></summary>

The appropriate implementation in this case is `naive_bayes`.`GaussianNB()`. Fit this model to your training data and predict on your test data.
</details>

<details>
<summary><h4><strong>Hint 3</strong></h4></summary>

Call `fit()`and pass your training feature set and target variable. Then call `predict()` on your test feature set.
</details>

## Step 4: Results and evaluation


### Leverage metrics to evaluate your model's performance

To evaluate the data yielded from your model, you can leverage a series of metrics and evaluation techniques from scikit-learn by examining the actual observed values in the test set relative to your model's prediction. Specifically, print the accuracy score, precision score, recall score, and f1 score associated with your test data and predicted values.

In [13]:
# Print your accuracy score.

### YOUR CODE HERE ###
print(metrics.accuracy_score(y_test, y_pred))


# Print your precision score.

### YOUR CODE HERE ###
print(metrics.precision_score(y_test, y_pred))


# Print your recall score.

### YOUR CODE HERE ###
print(metrics.recall_score(y_test, y_pred))


# Print your f1 score.

### YOUR CODE HERE ###
print(metrics.f1_score(y_test, y_pred))


0.6268656716417911
0.7785234899328859
0.5576923076923077
0.6498599439775911


<details>
<summary><h4><strong>Hint 1</strong></h4></summary>

Refer to [the content about model evaluation](https://www.coursera.org/learn/the-nuts-and-bolts-of-machine-learning/lecture/EITmV/key-evaluation-metrics-for-classification-models) for detail on these metrics.
</details>

<details>
<summary><h4><strong>Hint 2</strong></h4></summary>

The `metrics` module in `sklearn` has a function for computing each of these metrics.
</details>

<details>
<summary><h4><strong>Hint 3</strong></h4></summary>

Call `accuracy_score()`, `precision_score()`, `recall_score()`, and `f1_score()`, passing `y_test`, and `y_pred` into each function.
</details>

**Question:** What is the accuracy score for your model, and what does this tell you about the success of the model's performance?



62.69%. The model is performing decently good at predicting positives.

**Question:** Can you evaluate the success of your model by using the accuracy score exclusively?


No. To properly understand the performance of the model, especially classification model, we need other scores such as precision, recall and f1 scores too.

**Question:** What are the precision and recall scores for your model, and what do they mean? Is one of these scores more accurate than the other?


Precision score = 0.7785 and Recall score = 0.5576. Both these scores tell us how many Positives are correctly predicted from all the predicted postivies or all the actual postives.

**Question:** What is the F1 score of your model, and what does this score mean?

F1 score is a hybrid score. It is calculated using both Precission and Recall. F1 score of my model is 0.6498. This tells that the model is performing decently good at classifying but it's far from perfect. The perfect f1 score would be greater than or equal to 0.95.

### Gain clarity with the confusion matrix

Recall that a confusion matrix is a graphic that shows your model's true and false positives and negatives. It helps to create a visual representation of the components feeding into the metrics.

Create a confusion matrix based on your predicted values for the test set.

In [14]:
# Construct and display your confusion matrix.
import matplotlib.pyplot as plt
# Construct the confusion matrix for your predicted and test values.

### YOUR CODE HERE ###
confusion_matrix = metrics.confusion_matrix(y_test, y_pred)


# Create the display for your confusion matrix.

### YOUR CODE HERE ###
disp = metrics.ConfusionMatrixDisplay(confusion_matrix, )


# Plot the visual in-line.

### YOUR CODE HERE ###
disp.plot()
plt.show()

TypeError: __init__() missing 1 required positional argument: 'display_labels'

<details>
<summary><h4><strong>Hint 1</strong></h4></summary>

The `metrics` module has functions to create a confusion matrix.
</details>

<details>
<summary><h4><strong>Hint 2</strong></h4></summary>

Call `confusion_matrix`, passing in `y_test` and `y_pred`. Then, utilize `ConfusionMatrixDisplay()` to display your confusion matrix.
</details>

**Question:** What do you notice when observing your confusion matrix, and does this correlate to any of your other calculations?


[Write your response here. Double-click (or enter) to edit.]

## Considerations

**What are some key takeaways that you learned from this lab?**

[Write your response here. Double-click (or enter) to edit.]


**How would you present your results to your team?**

[Write your response here. Double-click (or enter) to edit.]


**How would you summarize your findings to stakeholders?**

[Write your response here. Double-click (or enter) to edit.]

